### PR TITLE
Serial provider class

### PR DESCRIPTION
**Pull request by CeQadrat**
ColdFilm parser nearly complite

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,3 +1,65 @@
...
+                    body.push(chunk);
+                })
+                .on('end', () => {
+                    body = Buffer.concat(body).toString();
+                    let data = parser(body);
+                    resolve(data);
+                })
+        });
+
+        request.on('error', (err) => {
```
Comment (radion-chernyakov) [inline app.js]: You can pass `reject` as collback here
`request.on('error', reject);`
also if arrow function's body have only one line it's better to write just `err => reject(err)`
https://github.com/airbnb/javascript#arrow-functions

File: app.js [hunk 2]
```diff
@@ -1,3 +1,65 @@
...
+        let options = {
+            hostname: 'coldfilm.ru',
+            port: 80,
+            method: 'GET',
+            path
+        };
+        let request = http.request(options, (res) => {
+            let body = [];
+            res
+                .on('data', (chunk) => {
```
Comment (radion-chernyakov) [inline app.js]: same here

File: app.js [hunk 3]
```diff
@@ -1,3 +1,65 @@
...
+                    resolve(data);
+                })
+        });
+
+        request.on('error', (err) => {
+            reject(err);
+        });
+
+        request.end();
+
```
Comment (radion-chernyakov) [inline app.js]: unnecessary blank line

File: app.js [hunk 4]
```diff
@@ -1,3 +1,65 @@
...
+    });
+}
+let name = 'Мистер робот';
+
+function* getSerial() {
+    let path = '/search/?q=' + utf8.encode(name).split(' ').join('+') + ';t=0;p=1;md=news';
```
Comment (radion-chernyakov) [inline app.js]: make separate function that will generate path

File: app.js [hunk 5]
```diff
@@ -1,3 +1,65 @@
...
+    let series = [];
+    while (true) {
```
Comment (radion-chernyakov) [inline app.js]: i think that you can make loop condition better :)
don't forget that there is not only `while() {}`
`do {} wile()` exist too

File: app.js [hunk 6]
```diff
@@ -1,3 +1,65 @@
...
+        let data = yield getRequest(path,coldFilmParsers.searchParser);
```
Comment (radion-chernyakov) [inline app.js]: try to avoid recursion

File: app.js [hunk 7]
```diff
@@ -1,3 +1,65 @@
...
+        path = data.nextPageLink;
+        series = series.concat(data.series);
+        if (!path) break;
+    }
+    return series;
+}
+
+co(getSerial())
+    .then((series) => {
+        console.log('Series num = ', series.length);
```
Comment (radion-chernyakov) [inline app.js]: looks like debug logs :)

File: app.js [hunk 8]
```diff
@@ -1,3 +1,65 @@
...
+        series.sort((s1,s2) => {
+            if(s1.date > s2.date) return 1;
+            if(s1.date < s2.date) return -1;
+        });
+        console.log(series);
```
Comment (radion-chernyakov) [inline app.js]: same

File: app.js [hunk 9]
```diff
@@ -1,3 +1,65 @@
...
+    })
+    .catch(err => console.error(err));
+
+getRequest('/news/mister_robot_1_sezon_10_serija_smotret_onlajn/2015-09-03-2677',coldFilmParsers.seriesParser)
```
Comment (radion-chernyakov) [inline app.js]: don't use hardcoded strings :)

Comment (radion-chernyakov) [reply app.js]: also u miss ` `
`foo('bar`, bar);`
not `foo('bar',baz);`

File: app.js [hunk 10]
```diff
@@ -1,3 +1,65 @@
...
+    .then((data) => {
+        console.log('Series: ');
+        console.log(data);
+    })
+    .catch(err => console.error(err));
```
Comment (radion-chernyakov) [inline app.js]: don't forget to add blank line at the end of file

File: coldFilmParsers.js [hunk 11]
```diff
@@ -0,0 +1,5 @@
+const searchParser = require('./coldFilmSearchParser');
+const seriesParser = require('./coldFilmSeriesParser');
+
+module.exports.seriesParser = seriesParser;
+module.exports.searchParser = searchParser;
```
Comment (radion-chernyakov) [inline coldFilmParsers.js:5]: rly?
module that import 2 modules and export them? :D

File: coldFilmSearchParser.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
+const htmlparser = require('htmlparser2');
+
```
Comment (radion-chernyakov) [inline coldFilmSearchParser.js:7]: too long function
it's hard to read it
and will be hard to maintain
don't forget to make sub-functions with good names

File: coldFilmSearchParser.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+function coldFilmSearchParser(str){
```
Comment (radion-chernyakov) [inline coldFilmSearchParser.js]: that meas `str`
give good name to this variable

File: coldFilmSearchParser.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    let parseLink = false;
```
Comment (radion-chernyakov) [inline coldFilmSearchParser.js]: is this variable will be `bool` for all time during function execution?
if it's true rename it
now it's sounds like in this variable will be `sting` or something like this, but not `bool`

File: coldFilmSearchParser.js (hunk 12 above)
Comment (radion-chernyakov) [reply coldFilmSearchParser.js:7]: also not forget to add new lines that will separate some statements

File: coldFilmSerialProvider.js [hunk 15]
```diff
@@ -0,0 +1,11 @@
+class ColdFilmSP{
+    constructor(serialName){
+        this.serialName = serialName;
+    }
+    get getSeries(){
+        return function* gen(){
```
Comment (radion-chernyakov) [inline coldFilmSerialProvider.js]: very useful generator :D

File: coldFilmSeriesParser.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
+const htmlparser = require("htmlparser2");
```
Comment (radion-chernyakov) [inline coldFilmSeriesParser.js]: same issues as for `coldFilmSearchParser`

File: app.js (hunk 6 above)
Comment (CeQadrat) [reply app.js]: It's not recursion. Function `getRequest` doesn't call itself :)

File: app.js [hunk 17]
```diff
@@ -1,3 +1,63 @@
...
+        let options = {
+            hostname: 'coldfilm.ru',
+            port: 80,
+            method: 'GET',
+            path
+        };
+        let request = http.request(options, (res) => {
+            let body = [];
+            res
+                .on('data', chunk => body.push(chunk))
```
Comment (radion-chernyakov) [inline app.js]: why still not `.on('data', body.push)`?

File: app.js [hunk 18]
```diff
@@ -1,3 +1,63 @@
...
+
+function* getSerial() {
+    let path = generatePath(name);
+    let series = [];
+    do{
+        let body = yield getRequest(path);
+        let data = coldFilmParsers.searchParser(body);
+        path = data.nextPageLink;
+        series = series.concat(data.series);
+    }while(path);
```
Comment (radion-chernyakov) [inline app.js]: miss probiel
yep, i tied to write in English :)

File: app.js [hunk 19]
```diff
@@ -1,3 +1,63 @@
-/**
- * Created by Dima on 10.09.2016.
- */
+const http = require('http');
+const coldFilmParsers = require('./coldFilmParsers');
+const utf8 = require('utf8');
+const co = require('co');
+
+function generatePath(name){
```
Comment (radion-chernyakov) [inline app.js]: looks good :+1:

File: app.js [hunk 20]
```diff
@@ -1,3 +1,63 @@
...
+        console.log(series);
+    })
+    .catch(err => console.error(err));
+
+getRequest('/news/mister_robot_1_sezon_10_serija_smotret_onlajn/2015-09-03-2677')
+    .then((data) => {
+        console.log('Series: ');
+        console.log(coldFilmParsers.seriesParser(data));
+    })
+    .catch(err => console.error(err));
```
Comment (radion-chernyakov) [inline app.js]: why here don't use `.catch(consol.error)`?

File: coldFilmSearchParser.js [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+    }, {decodeEntities: true});
+    parser.write(parseString);
+    parser.end();
+    return {
+        series,
+        nextPageLink
+    };
+}
+
+module.exports = coldFilmSearchParser;
```
Comment (radion-chernyakov) [inline coldFilmSearchParser.js:72]: it's better to move `exports` statement to the top of file
when you look at file you don't want see some backstage logic first, you want to see that is the interface of this module

File: coldFilmSearchParser.js [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+    let enableParseDate = false;
+    let enableParseNextPage = false;
+
+    let nextPageLink = '';
+    let serialTitle = '';
+    let series = [];
+    let serial = {};
+
+    let parser = new htmlParser.Parser({
+        onopentag: (name, attribs) => {
```
Comment (radion-chernyakov) [inline coldFilmSearchParser.js:14]: make separate functions `function onOpenTagCallback(name, attrs)` and pass it here
same for other callbacks

Comment (radion-chernyakov) [reply coldFilmSearchParser.js:14]: and same for other parser module

Comment (radion-chernyakov) [reply coldFilmSearchParser.js:14]: coz now it's looks ugly

File: app.js (hunk 17 above)
Comment (CeQadrat) [reply app.js]: This doesn't work
I don't understand why

**Pull request closed without merging**